### ISSUE TITLE
Database growth - and how to react

### ISSUE THREAD

**Issue by Try2Code** [buildbot/buildbot#7244]
Hi!
I am running a 3.x buildbot server with a postgresql database backend (bb-3.6/python-3.10/pgsql-10.23) since 2020.

My problem is, that the database is constantly growing at an unexpected rate. I already raised the disk space from 50 to 100GB in this spring, but now I am again close to the limit. It is clear, that the database must grow, but its rate needs extra care IMO.

The current server has done about 82k builds, leading to a database size of 85GB. The  I regularly run the postgres' `vacuumdb` for the data base, but with close-to-zero  size effects. My target project needs a testing infrastructure for the upcoming years, so I need a strategy for this growth. Unfortunately I could not find anything in the official documentation, yet. 

1. Does anybody have similar issues?
2. Are size and growth rate  I have expectable?
4. Is there a method to delete information about out-dated builds? builds from the last 12 months would be totally sufficient for the project.

Any hints are appreciated :pray:

**Comment by p12tic** [comment]
85GB for 82k of builds is 1MB per build. Could it be that you're not cleaning old logs? Check out https://docs.buildbot.net/latest/manual/configuration/configurators.html#janitorconfigurator

**Comment by Try2Code** [comment]
Thx so much @p12tic ! I dont clean old information at all. Never knew how to do it. This looks perfect

I expected something like this to appear in the chapter on deployment (https://docs.buildbot.net/latest/manual/deploy.html)

**Comment by Try2Code** [comment]
I added this to the config

```python
# DB clean up {{{
# delete logs older than half a year
c['configurators'] = [util.JanitorConfigurator(
    logHorizon=timedelta(weeks=30),
    hour=23,
    dayOfWeek=1
)] #}}}
```

but the database size did not get smaller :shrug:

**Comment by p12tic** [comment]
@Try2Code Did you try to run VACUUM on the tables in question? https://www.postgresql.org/docs/current/sql-vacuum.html

**Comment by Try2Code** [comment]
@p12tic yes, I do the vacuum-ing one or twice per year. Usually it is the first thing I do, when the harddisk reaches 95% . It never was effective. Which makes sense because I did not use `JanitorConfigurator()` before. But also after setting it to 10months the vacuuming does nothing. the interesting aspect it, that I does not grow beyond 95%. I have no idea, what postgres does internally. I think I will have to increase the diskspace once more and check what is happening. 

Regarding the `JanitorConfigurator()`: Is it designed to limit the database size?

**Comment by p12tic** [comment]
> Regarding the JanitorConfigurator(): Is it designed to limit the database size?

No, it just cleans old logs and similar information. If Buildbot is more active, it would result in more information stored in the log horizon and thus the database size would be larger.

I've checked postgres documentation - one needs to run VACUUM FULL to actually retrieve space back. See:

https://www.postgresql.org/docs/current/routine-vacuuming.html

> The standard form of VACUUM removes dead row versions in tables and indexes and marks the space available for future reuse. However, it will not return the space to the operating system, except in the special case where one or more pages at the end of a table become entirely free and an exclusive table lock can be easily obtained. In contrast, VACUUM FULL actively compacts tables by writing a complete new version of the table file with no dead space. This minimizes the size of the table, but can take a long time. It also requires extra disk space for the new copy of the table, until the operation completes.

**Comment by Try2Code** [comment]
thx a lot for the tip - and merry christmas :partying_face: 

```
vacuumdb: vacuuming of database "testing" failed: ERROR:  could not extend file "base/16396/21405183.3": No space left on device
```

I guess  I need to raise the diskspace so that complete rewrites of tables become possible again.

so in essence there is no configuration to tell the server to throw away any information older that a certain age?

**Comment by p12tic** [comment]
> so in essence there is no configuration to tell the server to throw away any information older that a certain age?

You could configure postgres to use table partitioning on the largest tables. This way you could drop old data by dropping child tables, which is much less expensive. https://github.com/pgpartman/pg_partman may help to automate some of the database administration overhead.

**Comment by Try2Code** [comment]
partitioning adds considerably to maintenance cost IMO. So I went with the _full vacuum_  after adding ur suggestions regarding the `JanitorConfigurator()`. This I configured to have a logHorizon of 30 weeks (the service started about 2 years ago, but with much less builders compared to now):  **Database size got reduced from 85GB to 46GB!**

Taken into account the non-linear increase of builders on the server over its lifetime this is a very good result for me. This should make the database handling manageable in the future. With possible technology changes in mind the server might have to  run for the next 5-10 years. Now it seems to be finally doable.

Thx again @p12tic for taking the time and digging into postgres - couldn't have done it without you :pray:

happy new year :balloon:
